### DISCUSSION TITLE
Register a controller to be the default one?

### DISCUSSION THREAD

**Discussion by lucian-torje-siemens** [eclipse-edc/Connector#1690] [Q&A]
Hello,
Replacing the existing controller with a custom one leads to an error - was this done on purpose and if yes what was the reason?

Would it be a good idea to be able to replace the controller or even better to refactor the code so that each controller delegates processing to an injected service (replaceable) ?  (maybe start with MultipartController)

![image](https://user-images.githubusercontent.com/102717569/178932483-7298ccba-e9a3-4251-9491-4f244f17961b.png)

**Comment by juliapampus** [comment]
What do your mean with "leads to _an_ error"? This is a really vague statement. Producing errors is never intended. 

What is your use case and why do you want to replace the current IDS implementation (as this is what this extension is providing)?

**Reply by lucian-torje-siemens** [reply]
An extension trying to register the same endpoint twice would get an error like the one below:

![image](https://user-images.githubusercontent.com/102717569/178964320-7850aa08-3f96-4d16-aea0-16b1ac444c25.png)

The intension is to extend the functionality provided by a controller - in our specific case, is to enhance the claim token with an extra property, but this functionality would allow anyone else to use it for other purposes (hence the question if it was designed like this on purpose).

**Reply by lucian-torje-siemens** [reply]
This happens due to:

![image](https://user-images.githubusercontent.com/102717569/178965686-c3521d58-f43e-4ff6-ad63-86017ed2c022.png)

**Reply by ndr-brt** [reply]
@lucian-torje-siemens if I understood this correctly and the need is to add a claim to the JWT token you would only need to register a `JwtDecorator` to the `Oauth2JwtDecoratorRegistry` service.

**Reply by lucian-torje-siemens** [reply]
I was thinking to use that too (DapsJwtDecorator is using it as well) - from my findings it does not look that those properties get set back into the claimtoken (please correct me if I'm wrong) - what I would need is something like this https://github.com/mindsphere/DataSpaceConnector-Fork/pull/20/files#diff-40e48cf0c24ef84a069be0592986f122fcafcc4f1e6b84df1ad37dc9b61076ce

![image](https://user-images.githubusercontent.com/102717569/178984834-c473e461-d7e8-4d2c-baed-d5076484d513.png)

**Reply by ndr-brt** [reply]
The claims are decorated in the IdsSender part before sending the ids-multipart request, not in the Controller (where the request is received)

So your intent is to add a claim based on an header in the multipart controller based on an header. Just out of curiosity, who's gonna make that call with that ten header? another EDC right? If so the JWTDecorator is what you are looking for because it adds a claim before signing the token, then the token is put in the multipart header and so it will be extracted by the MultipartController when it will receive the call.

**Reply by lucian-torje-siemens** [reply]
Thank you @ndr-brt for the hint - I will test it.

**Reply by lucian-torje-siemens** [reply]
Just checked with daps - the steps are below:
- the first EDC creates the token and adds the extra claim (now using an JWTDecorator)
- the additional claims are only send to daps and not saved afterwards (so my tenant claim is removed) --> my expectation here would be that the extra claims are keep as additional and sent to the second EDC like here [https://github.com/mindsphere/DataSpaceConnector-Fork/pull/20/commits/b22e9f7413e4a828e89d168414ac744aba534349](https://github.com/mindsphere/DataSpaceConnector-Fork/pull/20/commits/b22e9f7413e4a828e89d168414ac744aba534349)
- the  token is sent to the second EDC without the tenant claim

**Reply by ndr-brt** [reply]
You could configure the DAPS server to return the `ten` claims in the token

**Reply by lucian-torje-siemens** [reply]
We do not have access to the DAPS server; your hint make me realize that I need to recheck the JWT token itself and it looks like the JWT token contains now the new attribute tenant added by the my JWTDecorator implementation.
